src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -217,6 +244,39 @@ private IReadOnlyList<SemanticRange> CombineSemanticRanges(params IReadOnlyList<
             return result;
         }
 
+        private async Task<SemanticTokens> GetMatchingCSharpResponseAsync(
+            TextDocumentIdentifier textDocumentIdentifier,
+            long? documentVersion,
+            CancellationToken cancellationToken)
+        {
+            var parameter = new SemanticTokensParams
+            {
+                TextDocument = textDocumentIdentifier,
+            };
+
+            ProvideSemanticTokensResponse csharpResponse;
+            for (var i = 0; i < 3; i++)
+            {
+                var request = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorProvideSemanticTokensEndpoint, parameter);
```

Comment:
Just double checking if this is right. Feels odd since it's an async operation to enqueue the request but then ontop of that it doesn't take a cancellation token 😆

Reply:
The new way that O# works SendRequestAsync does very little, all of the going out to the LSP and retrieving results happens in `request.Returning` below. I assume that's why SendRequestAsync doesn't take a CancellationToken.

Reply:
@david-driscoll Why does the "request building" API here have to be async out of curiosity?